### PR TITLE
[2.6] Fix for #7068: EntityManager::find() with pessimistic lock should check for transaction

### LINKED ISSUE #7068: EntityManager find with pessimistic lock does not check for transaction when cached
When a pessimistic lock mode is passed to EntityManager::find it normally throws a TransactionRequiredException if no transaction is currently active.

However if the entity is currently in the identity map, it will refresh the entity from the database and return it but not check if a transaction is running. The refresh query has 'FOR UPDATE' (in case of pessimistic write) but this has no effect since there is no transaction.

To reproduce:
```php
// $em is EntityManager, Order entity can be any entity
$em->find(Order::class, 1); // Load order id 1 in entity map
$em->find(Order::class, 1, LockMode::PESSIMISTIC_WRITE); // does not throw exception
$em->find(Order::class, 2, LockMode::PESSIMISTIC_WRITE); // throws TransactionRequiredException 
```
Tested with doctrine 2.5.14 but latest code seems to have the same issue.

### PR DESCRIPTION
**Pull request by Majkl578**
Rebased and adjusted #7075 for 2.6.
Fixes #7068.

### CODE REVIEW

File: lib/Doctrine/ORM/EntityManager.php [hunk 1]
```diff
@@ -915,4 +912,26 @@ public function hasFilters()
...
         return null !== $this->filterCollection;
     }
+
+    /**
+     * @param int $lockMode
+     * @param ClassMetadata $class
+     * @throws OptimisticLockException
+     * @throws TransactionRequiredException
+     */
+    private function checkLockRequirements(int $lockMode, ClassMetadata $class): void
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: Nit: we use a space between `)` and `:`. Not important though.

**Review by Ocramius — APPROVED** [review]

**Comment by Ocramius** [comment]
Thanks @Majkl578!

File: lib/Doctrine/ORM/EntityManager.php (hunk 1 above)
Comment (Majkl578) [reply lib/Doctrine/ORM/EntityManager.php]: Hmm, leftover from original PR.

**Comment by timdev** [comment]
Why the intentional fallthrough here?  It causes a BC break when using optimistic locking outside of a tx.  I'm probably missing something, but it just bit me.

**Comment by Majkl578** [comment]
I think you are right, the transaction shouldn't be required for an optimistic lock ([documentation examples sagree](https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/transactions-and-concurrency.html#optimistic-locking)), at least not in 2.x. Do you want to submit a PR? :+1:

**Comment by timdev** [comment]
My pleasure: https://github.com/doctrine/doctrine2/pull/7367

**Pull request merged**